File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int res = 0;
        int n = prices.length;
        int max = 0;
        for(int i=n-1;i>=0;i--){
            max = Math.max(prices[i], max);
            res = Math.max(res, max-prices[i]);
        }
        return res;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0
        curr_max = 0
        for i in range(len(prices)-1, -1, -1):
            curr_max = max(curr_max, prices[i])
            res = max(res, curr_max - prices[i])
        return res
            
            
        